### PR TITLE
Ntgl 404 debug timetable sharing

### PR DESCRIPTION
**Pull request by Rayahhhmed**
Fixed the old method of parsing timetable sharing by making new method more robust to user cases 🦖

### CODE REVIEW

**Review by manhualu — APPROVED** [review]

**Comment by Rayahhhmed** [comment]
Very good work

File: client/src/components/ctfs/TimetablesFlag.tsx [hunk 1]
```diff
@@ -29,44 +29,69 @@ const TimetablesFlag: React.FC = () => {
...
+    // ARTS1631: number[];
+    // COMM1140: number[];
+    // COMP1511: number[];
+  }
+  const isValidWin = (initData: CourseMap[]): boolean => {
+    if (initData.length !== 3) return false;
+    initData.sort((a, b) => a.timetableMapped.length - b.timetableMapped.length);
+    // Greedy matching
+
+    let timetablesIdsUsed: number[] = [];
```
Comment (zax-xyz) [inline client/src/components/ctfs/TimetablesFlag.tsx:46]: comment should be moved down to be with the relevant block of code

File: client/src/components/ctfs/TimetablesFlag.tsx [hunk 2]
```diff
@@ -29,44 +29,69 @@ const TimetablesFlag: React.FC = () => {
...
 
-  const flag = 'level{using_notangles_is_ez}';
+  const flag = 'levelup{using_notangles_is_ez}';
+  const coursesOfInterest: string[] = ['ARTS1631', 'COMM1140', 'COMP1511'];
+  interface CourseMap {
+    name: string;
+    timetableMapped: number[];
+    // ARTS1631: number[];
+    // COMM1140: number[];
+    // COMP1511: number[];
```
Comment (zax-xyz) [inline client/src/components/ctfs/TimetablesFlag.tsx:39]: remove commented code from final submission pls

File: client/src/components/ctfs/TimetablesFlag.tsx [hunk 3]
```diff
@@ -29,44 +29,69 @@ const TimetablesFlag: React.FC = () => {
...
-            doesComm = true;
-          }
-          if (course.code === 'COMP1511') {
-            console.log('does comp');
-            doesComp = true;
-          }
-          if (doesArts && doesComm && doesComp) {
-            setFlagDialogOpen(true);
+
+      timetables.forEach((timetable: any, idx: number) => {
```
Comment (zax-xyz) [inline client/src/components/ctfs/TimetablesFlag.tsx:71]: can we use a more specific type here than `any`

File: client/src/components/ctfs/TimetablesFlag.tsx [hunk 4]
```diff
@@ -29,44 +29,69 @@ const TimetablesFlag: React.FC = () => {
...
+      setFlagDialogOpen(true);
     }
   };
 
   const handleDialogClose = () => {
     setFlagDialogOpen(false);
     setFlagShown(true);
   };
-
+  let timetables = storage.get('timetables');
```
Comment (zax-xyz) [inline client/src/components/ctfs/TimetablesFlag.tsx:91]: why is this not const? this variable doesn't seem to change

File: client/src/components/ctfs/TimetablesFlag.tsx [hunk 5]
```diff
@@ -29,44 +29,69 @@ const TimetablesFlag: React.FC = () => {
...
-  const flag = 'level{using_notangles_is_ez}';
+  const flag = 'levelup{using_notangles_is_ez}';
+  const coursesOfInterest: string[] = ['ARTS1631', 'COMM1140', 'COMP1511'];
+  interface CourseMap {
+    name: string;
+    timetableMapped: number[];
+    // ARTS1631: number[];
+    // COMM1140: number[];
+    // COMP1511: number[];
+  }
```
Comment (zax-xyz) [inline client/src/components/ctfs/TimetablesFlag.tsx:40]: thoughts on having this interface outside the function?

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 1 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:46]: Yeh tru was kinda greedy with that one

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 2 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:39]: No v imp comment

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 3 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:71]: Akshually timetable is so doomed we don’t have strong typings for it

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 4 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:91]: Good pick

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 3 above)
Comment (zax-xyz) [reply client/src/components/ctfs/TimetablesFlag.tsx:71]: doomed codebase tbh

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 5 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:40]: Nothing else would ever use this sesspool of an interface tho

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 3 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:71]: Yeh wanna rewrite it next year subcomm is open

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 5 above)
Comment (zax-xyz) [reply client/src/components/ctfs/TimetablesFlag.tsx:40]: this is the first time i've seen an interface defined inside a function so seemed weird, and i didn't see any other instance of this in the notangles codebase so not sure if this is established style

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 3 above)
Comment (zax-xyz) [reply client/src/components/ctfs/TimetablesFlag.tsx:71]: yeah sure let's rewrite it in rust

File: client/src/components/ctfs/TimetablesFlag.tsx (hunk 5 above)
Comment (Rayahhhmed) [reply client/src/components/ctfs/TimetablesFlag.tsx:40]: It just works so we ball

**Pull request merged**